app/controllers/application_controller.rb

```diff
@@ -5,14 +5,10 @@ class ApplicationController < ActionController::Base
   before_action :set_locale
 
   def set_locale
-    begin
-      if params[:locale] != nil
-        cookies.permanent[:locale] = params[:locale]
-      end
-      I18n.locale = cookies[:locale] || read_lang_header || I18n.default_locale
-    rescue I18n::InvalidLocale
-      I18n.locale = I18n.default_locale
-    end
+    cookies.permanent[:locale] = params[:locale] if params[:locale].present?
+    I18n.locale = cookies[:locale] || current_user_locale.presence || read_lang_header || I18n.default_locale
```

Comment:
Hi @sillylogger,
 I think we don't need use `cookie` for this. In case `cookies` exist, when user change locale in profile page, `I18n.locale` will not be changed.